src/NumericalAlgorithms/LinearOperators/PartialDerivatives.tpp

```diff
@@ -7,25 +7,212 @@
 
 #include "DataStructures/DataBox/DataBoxTag.hpp"
 #include "DataStructures/DataBox/Prefixes.hpp"
+#include "DataStructures/DataVector.hpp"
 #include "DataStructures/Matrix.hpp"
 #include "DataStructures/Variables.hpp"
 #include "Domain/Mesh.hpp"
 #include "NumericalAlgorithms/LinearOperators/Transpose.hpp"
 #include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Algorithm.hpp"
 #include "Utilities/Blas.hpp"
+#include "Utilities/Gsl.hpp"
 #include "Utilities/MakeArray.hpp"
 #include "Utilities/StdArrayHelpers.hpp"
 
 namespace partial_derivatives_detail {
 template <size_t Dim, typename VariableTags, typename DerivativeTags>
 struct LogicalImpl;
+
+// This routine has been optimized to perform really well. The following
+// describes what optimizations were made.
+//
+// - Raw pointers to the logical derivatives are passed in so that in the case
+//   where `partial_derivatives` is not passed in the logical derivatives as an
+//   argument but needs to compute them can allocate the entire buffer for the
+//   logical partial derivatives at once.
+//
+// - The resultant Variables is a not_null pointer so that mutating compute
+//   items can be supported.
+//
+// - The storage indices into the inverse Jacobian are precomputed to avoid
+//   having to recompute them for each tensor component of `u`.
+//
+// - The DataVectors lhs and logical_du are used as non-owning DataVectors to be
+//   able to plug into the optimized expression templates. This requires a
+//   `const_cast` even though we will never change the `double*`.
+//
+// - Instead of looping over the Tensors directly using a `tmpl::for_each` we
+//   loop over the components using a raw pointer.
```

Comment:
Nope, you've not understood correctly... I'm not sure how to clarify because the old code literally looped over Tensors using a `tmpl::for_each`. Suggestions?

Reply:
How about: "Instead of looping over the Tensors in `du` directly using a `tmpl::for_each`, we loop using a raw pointer."